File: octave/export_result.m
% Hydrophone Deconvolution - Multi-language Implementation
%
% Based on Tutorial-Deconvolution by Weber & Wilkens (2023)
% DOI: 10.5281/zenodo.10079801
% License: CC BY 4.0

addpath(fullfile(fileparts(mfilename('fullpath'))));

% Load test data
measured_signal = load('../test-data/measured_signal.csv')';
freq_real = load('../test-data/freq_response_real.csv')';
freq_imag = load('../test-data/freq_response_imag.csv')';
frequency_response = freq_real + 1i * freq_imag;

% Deconvolve
result = deconvolution.deconvolve_without_uncertainty(measured_signal, frequency_response, 1e7);

% Save result
results_dir = '../validation/results';
if ~exist(results_dir, 'dir')
    mkdir(results_dir);
end
dlmwrite(fullfile(results_dir, 'octave_result.csv'), result', '%.18e');
fprintf('Octave result exported: %d samples\n', length(result));

% Pulse parameters
signal_uncertainty = load('../test-data/signal_uncertainty.csv')';
n = length(result);
sampling_rate = 1e7;
time = (0:n-1) / sampling_rate;
pp = deconvolution.pulse_parameters(time, result', signal_uncertainty);
fid = fopen(fullfile(results_dir, 'octave_pulse_params.csv'), 'w');
fprintf(fid, '%.18e,%.18e,%.18e,%.18e,%.18e,%.18e\n', ...
    pp.pc_value, pp.pc_uncertainty, pp.pr_value, pp.pr_uncertainty, ...
    pp.ppsi_value, pp.ppsi_uncertainty);
fclose(fid);
fprintf('Octave pulse parameters exported\n');

File: cpp/tests/export_result.cpp
/*
 * Hydrophone Deconvolution - Multi-language Implementation
 *
 * Based on Tutorial-Deconvolution by Weber & Wilkens (2023)
 * DOI: 10.5281/zenodo.10079801
 * License: CC BY 4.0
 */

#include "deconvolution.hpp"
#include <fstream>
#include <iostream>
#include <iomanip>
#include <string>
#include <filesystem>

Eigen::VectorXd load_csv(const std::string& path) {
    std::ifstream file(path);
    std::vector<double> values;
    double val;
    while (file >> val) {
        values.push_back(val);
    }
    Eigen::VectorXd result(values.size());
    for (size_t i = 0; i < values.size(); ++i) {
        result(i) = values[i];
    }
    return result;
}

int main() {
    std::string data_dir = "../../test-data/";
    std::string results_dir = "../../validation/results/";
    std::filesystem::create_directories(results_dir);

    auto measured = load_csv(data_dir + "measured_signal.csv");
    auto freq_real = load_csv(data_dir + "freq_response_real.csv");
    auto freq_imag = load_csv(data_dir + "freq_response_imag.csv");

    Eigen::VectorXcd freq_response(freq_real.size());
    for (int i = 0; i < freq_real.size(); ++i) {
        freq_response(i) = std::complex<double>(freq_real(i), freq_imag(i));
    }

    auto result = hydrophone::deconvolve_without_uncertainty(measured, freq_response, 1e7);

    std::ofstream out(results_dir + "cpp_result.csv");
    out << std::setprecision(18) << std::scientific;
    for (int i = 0; i < result.size(); ++i) {
        out << result(i) << "\n";
    }

    std::cout << "C++ result exported: " << result.size() << " samples" << std::endl;
    return 0;
}
